pkg/internal/cyberark/identity/identity.go

```diff
@@ -0,0 +1,440 @@
+	}
+
+	defer httpResponse.Body.Close()
+
+	if httpResponse.StatusCode != 200 {
+		return response, fmt.Errorf("got unexpected status code %s from request to start authentication in CyberArk Identity API", httpResponse.Status)
+	}
+
+	startAuthResponse := startAuthenticationResponseBody{}
+
+	err = json.NewDecoder(io.LimitReader(httpResponse.Body, maxStartAuthenticationBodySize)).Decode(&startAuthResponse)
+	if err != nil {
+		if err == io.ErrUnexpectedEOF {
+			return response, fmt.Errorf("rejecting JSON response from server as it was too large or was truncated")
+		}
+
+		return response, fmt.Errorf("failed to parse JSON from otherwise successful request to start authentication: %s", err)
+	}
+
+	if !startAuthResponse.Success {
+		return response, fmt.Errorf("got a failure response from request to start authentication: message=%q, error=%q", startAuthResponse.Message, startAuthResponse.ErrorID)
+	}
+
+	logger.V(logs.Debug).Info("made successful request to StartAuthentication", "summary", startAuthResponse.Result.Summary)
+
+	if startAuthResponse.Result.Summary != SummaryNewPackage {
+		// This means we can't respond to whatever summary the server sent.
+		// The best thing to do is try and find a challenge we can solve anyway.
+		klog.FromContext(ctx).Info("got an unexpected Summary from StartAuthentication response; will attempt to complete a login challenge anyway", "summary", startAuthResponse.Result.Summary)
+	}
+
+	if len(startAuthResponse.Result.Challenges) == 0 {
+		return response, fmt.Errorf("got no valid challenges in response to start authentication; unable to log in")
+	}
+
+	mechanismID := ""
+
+	for i, challenge := range startAuthResponse.Result.Challenges {
+		logger.V(logs.Debug).Info("found a challenge", "idx", i, "mechanismCount", len(challenge.Mechanisms))
+
+		if len(challenge.Mechanisms) == 0 {
+			// presumably this shouldn't happen, but handle the case anyway
+			logger.Info("got no mechanisms for challenge from Identity server; skipping this challenge")
+			continue
+		}
+
+		for j, mechanism := range challenge.Mechanisms {
+			logger.V(logs.Debug).Info("found a mechanism in challenge", "idx", j, "enrolled", mechanism.Enrolled, "name", mechanism.Name)
+
+			if !mechanism.Enrolled || mechanism.Name != MechanismUsernamePassword {
+				continue
+			}
+
+			// got a username/password mechanism, so use it
+			mechanismID = mechanism.MechanismID
+			break
+		}
+	}
+
+	if mechanismID == "" {
+		klog.FromContext(ctx).Info("ctx", "response", startAuthResponse)
+		return response, errNoUPMechanism
+	}
+
+	response.Action = ActionAnswer
+	response.MechanismID = mechanismID
+	response.SessionID = startAuthResponse.Result.SessionID
+	response.TenantID = c.subdomain
+	response.PersistantLogin = true
+
+	return response, nil
+}
+
+func (c *Client) doAdvanceAuthentication(ctx context.Context, username string, requestBody *advanceAuthenticationRequestBody) error {
+	bodyJSON, err := json.Marshal(requestBody)
```

Comment:
nit: Seems like doStartAuthentication and doAdvanceAuthentication methods have quite a lot of common code. Does it make sense to extract common template methods?

Reply:
My instinct is that it's not worth extracting it right now - there are enough differences between the two that it would be a bit harder to read if I did that, IMO. If we had to add a third method which was similar I'd definitely start to consider though!